be/src/exec/workgroup/pipeline_executors.cpp

```diff
@@ -0,0 +1,181 @@
+    return fmt::format(
+            "([num_total_cores={}] [num_total_driver_threads={}] [num_total_scan_threads={}] "
+            "[num_total_connector_scan_threads={}] [enable_bind_cpus={}] [enable_cpu_borrowing={}])",
+            num_total_cores, num_total_driver_threads, num_total_scan_threads, num_total_connector_scan_threads,
+            enable_bind_cpus, enable_cpu_borrowing);
+}
+
+// ------------------------------------------------------------------------------------
+// PipelineExecutors
+// ------------------------------------------------------------------------------------
+
+PipelineExecutors::PipelineExecutors(const PipelineExecutorsConfig& conf, std::string name, CpuUtil::CpuIds cpuids,
+                                     std::vector<CpuUtil::CpuIds> borrowed_cpuids)
+        : _conf(conf),
+          _name(std::move(name)),
+          _cpuids(std::move(cpuids)),
+          _borrowed_cpu_ids(std::move(borrowed_cpuids)) {}
+
+PipelineExecutors::~PipelineExecutors() {
+    close();
+}
+
+std::string PipelineExecutors::to_string() const {
+    return fmt::format(
+            "([name={}] [num_driver_threads={}] [num_scan_threads={}] [num_connector_scan_threads={}] [cpuids={}] "
+            "[conf={}])",
+            _name, num_driver_threads(), num_scan_threads(), num_connector_scan_threads(), CpuUtil::to_string(_cpuids),
+            _conf.to_string());
+}
+
+Status PipelineExecutors::start() {
+    if (_stage >= Stage::STARTED) {
+        return Status::OK();
+    }
+    _stage = Stage::STARTED;
+
+    std::unique_ptr<ThreadPool> driver_executor_thread_pool;
+    RETURN_IF_ERROR(ThreadPoolBuilder("pip_exec_" + _name)
+                            .set_min_threads(0)
+                            .set_max_threads(num_driver_threads())
+                            .set_max_queue_size(1000)
+                            .set_idle_timeout(MonoDelta::FromMilliseconds(2000))
+                            .set_cpuids(_cpuids)
+                            .set_borrowed_cpuids(_borrowed_cpu_ids)
+                            .build(&driver_executor_thread_pool));
+    _driver_executor = std::make_unique<pipeline::GlobalDriverExecutor>(_name, std::move(driver_executor_thread_pool),
+                                                                        true, _cpuids);
+    _driver_executor->initialize(num_driver_threads());
+
+    std::unique_ptr<ThreadPool> scan_thread_pool;
+    RETURN_IF_ERROR(ThreadPoolBuilder("pip_scan_" + _name)
+                            .set_min_threads(0)
+                            .set_max_threads(num_scan_threads())
+                            .set_max_queue_size(1000)
+                            .set_idle_timeout(MonoDelta::FromMilliseconds(2000))
+                            .set_cpuids(_cpuids)
+                            .set_borrowed_cpuids(_borrowed_cpu_ids)
+                            .build(&scan_thread_pool));
+    _scan_executor = std::make_unique<ScanExecutor>(
```

Comment:
so each resource group has a PipelineExecutors, each PipelineExecutors has a _scan_executor, __connector_scan_executor and _driver_executor?

all the PipelineExecutors belonging to dedicated resource groups or common ones are kept in singleton WorkgroupManager?